nodewatcher/nodewatcher.py

```diff
@@ -154,35 +169,46 @@ def main():
     log.info("nodewatcher startup")
     instance_id = getInstanceId()
     hostname = getHostname()
-    region, asg_name, scheduler, proxy_config = getConfig(instance_id)
+    region, asg_name, scheduler, proxy_config, idle_time = getConfig(instance_id)
+
 
     s = loadSchedulerModule(scheduler)
 
+    if os.path.isfile(_NW_IDLETIME_FILE):
+        data = json.loads(open(_NW_IDLETIME_FILE).read())
+    else:
+        data = {'idle_time': 0}
+
+    atexit.register(saveIdleTime, data)
     while True:
         time.sleep(60)
-        ec2_conn = boto3.resource('ec2', region_name=region, config=proxy_config)
         asg_conn = boto3.client('autoscaling', region_name=region, config=proxy_config)
-        hour_percentile = getHourPercentile(instance_id, ec2_conn)
-        log.info('Percent of hour used: %d' % hour_percentile)
-
-        if hour_percentile < 95:
-            continue
 
         jobs = getJobs(s, hostname)
-        if jobs == True:
+        if jobs:
             log.info('Instance has active jobs.')
+            data['idle_time'] = 0
         else:
             if maintainSize(asg_name, asg_conn):
                 continue
             # avoid race condition by locking and verifying
             lockHost(s, hostname)
```

Comment:
I'd avoid the lock host operation when not strictly necessary.
The 'else' logic could be rephrased like:
```
  increment_idle_time
  if idle_time > scale-down-threshold
     lockHost()
     if not hasJobs() and not pendingJobs()
         selfTerminate()
     else
         unlockHost()
```

Reply:
I left the existing design as was before my changes.

Reply:
yep but in the previuos version if you were in this part of the code all the conditions to terminate the node were potentially satisfied (hourPercentile >= 95 && ! maintainSize) so we could lock the host, check for jobs and, if the case, terminate the instance.
In the new version, we've still to check if the passed idle_time "is enough" and only in this case it makes sense to disturb the scheduler to lock the host and check if new jobs were submitted in the meantime etc.